graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -480,25 +486,31 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
             else:
                 raise AssertionError(u'Output inside @fold scope did not add '
                                      u'"fold_innermost_scope" to context! '
-                                     u'Location: {}'.format(fold_scope_location))
+                                     u'Location: {}'.format(location))
 
         if in_topmost_optional_block:
             basic_blocks.append(blocks.EndOptional())
             del context['optional']
 
         # If we are currently evaluating a @fold vertex,
         # we didn't Traverse into it, so we don't need to backtrack out either.
-        # Alternatively, we don't backtrack if we've reached an @output_source.
+        # We also don't backtrack if we're currently inside a @fold scope, or
+        # if we've reached an @output_source.
         backtracking_required = (
-            (not fold_directive) and (not has_encountered_output_source(context)))
+            (not fold_directive) and
+            (not isinstance(location, FoldScopeLocation)) and
+            (not has_encountered_output_source(context))
+        )
```

Comment:
The issue with `all/any` is that when you pass them a tuple, the entire tuple has to be eagerly constructed before they can evaluate it. When they consume a generator, this isn't a problem since they never see the elements that aren't required.

Normally I don't care either way, though I think @lodrion has established that his preference is for `and/or` explicitly rather than `any/all`. But in this case, since we're calling functions, I'd much rather not call those functions if the result can be determined from the first element. This is why the clauses are ordered in rough order of increasing complexity.